content/0000-container-entrypoints/rfc.md

```diff
@@ -0,0 +1,247 @@
+technically.
+
+We also have some prior art to work from, since Tekton uses a mechanism very
+similar to this (albeit with an init container) to inject functionality into
+task containers.
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+To place the entrypoint binary in the container, we need to either copy files
+into it using, e.g., an init container and shared volume or have the files
+pre-allocated in, e.g., a PersistentVolumeClaim or ConfigMap.
+
+Tekton doesn't support init containers, but we could relatively easily use
+[workspaces](https://tekton.dev/docs/pipelines/workspaces/). An init Task would
+be a dependency for every other Task in the Pipeline. This introduces some
+complexity around scheduling; namely, it forces the entire PipelineRun to run on
+the same node using a process they call the affinity assistant. It also means
+that we can't use our own affinities in the future.
+
+Unfortunately, Knative Serving doesn't provide any useful functionality to run
+init containers or inject PVCs into pods. A [discussion in their issue
+tracker](https://github.com/knative/serving/issues/4579) points to an
+implementation by the Kubeflow folks that uses a mutating webhook to get what
+they (and we) want. Although I'm generally wary of bypassing APIs like this,
+we're already doing it for a number of other pod fields and it works well.
+
+Finally, etcd's configuration limits the size of ConfigMaps to 1MB, so while we
+could probably pull off injecting a small binary written in C or C++, we can't
+use any of the languages we're familiar with as a team.
+
+### What is the impact of not doing this?
+
+Supporting the features associated with our desired step execution flow would
+either fall to a more complex external management process or to the user (i.e.,
+forcing them to use a specific framework). Some features would be impossible;
+for example, exposing environment variables with secrets would require us to
+make unacceptable security compromises.
+
+### What specific risks are associated with this design?
+
+* We haven't fully vetted the PVC approach. For example, with GCE-PD we'll need
+  to provision regional disks, at which point each tenant will be restricted to
+  running in two zones instead of three. Since the PVCs will be distributed
+  across tenants we should still retain very high availability guarantees
+  barring other unforeseen issues.
+* There is a cultural risk that the entrypointer binary may feel like trusted
+  code. It is not. It runs directly in a user-controlled environment without any
+  separation from its delegate process. For example, the delegate process might
+  directly read or manipulate the memory of the entrypointer.
+* Conversely, if our own code is insecure, we introduce a potential attack
+  vector to users' containers.
+* We recognize that this may make compatibility with operating systems and
```

Comment:
We can largely avoid this by using Go, which cross-compiles quite easily to architectures the Linux kernel supports.